### PR TITLE
add HTTPS-enabled reverse proxy deployment example

### LINKED ISSUE #1167: Standard reverse proxy for HTTPS
Some users setting up their own deployment of Aerie desire a standalone reverse proxy to enable HTTPS between clients and Aerie components. This is especially important when using `CAM` authorization, since users need to pass their credentials through the UI. This is currently achieved for the development deployment (`aerie-dev.jpl.nasa.gov`) by using an AWS Application Load Balancer. This ticket should investigate and create an example deployment with a TLS-terminating reverse proxy in the `docker-compose` stack, like `nginx`.

### PR DESCRIPTION
**Pull request by skovati**
* **Tickets addressed:** Closes #1167 
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
**TL;DR**: New optional container in compose stack listens for HTTPS requests and forwards them as HTTP requests to internal containers. This enables encryption of traffic between clients and Aerie and reduces the number of exposed ports of an Aerie deployment.

Currently, the deployment we provide to customers is fairly barebones, and simply spins up the Aerie components with all ports exposed, and lets users hit `localhost:80` for the UI. This deployment is useful for local development purposes and quick testing, but a production Aerie deployment would likely want to encrypt traffic between the browser and Aerie components, using HTTPS, and be deployed under a specific domain name (e.g. `aerie-dev.jpl.nasa.gov`)

This PR adds a new optional component `aerie_proxy` to the docker compose stack, which runs an [`nginx`](https://nginx.org/en/) instance (a popular open-souce HTTP webserver and reverse proxy) in a "reverse proxy" configuration. This new container will listen on the provided domain for just the essential ports of an Aerie deployment (`443` for the UI, `8080` for Hasura, and `9000` for the gateway), create a TLS tunnel with connecting clients using the provided TLS certificates, which on the communicate with the client on these ports using TLS with the provided certificates, and then terminate TLS and proxy the traffic to internal containers within the network namespace provided by docker.

For clarification, this component doesn't really offer much utility when deploying locally for development, and is mostly useful as an example pattern for end-users to deploy Aerie with HTTPS.

An example client <-> server interaction with this new reverse proxy is as follows:

```mermaid
sequenceDiagram
    participant browser
    participant nginx
    participant hasura
    browser->>nginx: TLS handshake init (ClientHello)
    nginx->>browser: TLS handshake complete
    browser->>nginx: HTTPS request for aerie.test:8080/v1/graphql with JSON payload
    Note over nginx: Nginx strips TLS
    nginx->>hasura: Forward HTTP request to hasura:8080/v1/graphql
    hasura->>nginx: Query result HTTP response (JSON body)
    nginx->>browser: Query result HTTPS response (JSON body)
```

## Verification
The following manual steps were done to verify this new component works correctly:
- Generate self-signed TLS cert for testing:
  - `openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem -sha256 -days 365`
  - Move certs to `proxy` folder
- Edit `deployment/.env` to add new `AERIE_HOST` entry as `aerie.test`
- Edit `/etc/hosts` to add `127.0.0.1 aerie.test` as a fake DNS entry
- Run docker compose up with both the deployment yaml and the new proxy overlay
  -  `docker compose -f docker-compose.yml -f proxy/docker-compose-proxy.yml up -d --build`
- Visit `https://aerie.test` and confirm regular Aerie functionality.

## Documentation
Need to type up docs under the advanced deployment section of `aerie-docs`.

## Future work
Possibly continue this effort of creating patterns for common deployment needs (database backups, 3rd party identity providers, etc)

### CODE REVIEW

File: deployment/proxy/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from docker.io/nginx:alpine
```
Comment (cohansen) [inline deployment/proxy/Dockerfile]: Do we want to target a specific version here like we do for our other images?

File: deployment/proxy/docker-compose-proxy.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+     - "443:443"
+     - "8080:8080"
+     - "9000:9000"
+    environment:
+        NGINX_ENVSUBST_TEMPLATE_SUFFIX: ".template"
+        AERIE_HOST: ${AERIE_HOST}
+    networks:
+      - aerie_net
+  aerie_gateway:
+    ports: !reset []
```
Comment (cohansen) [inline deployment/proxy/docker-compose-proxy.yml:21]: What does `!reset []` do?

**Review by cohansen — APPROVED** [review]
Nice diagram 👀 .

Just a few questions, it all looks good to me. I'll run everything tomorrow to double check that your instructions are good.

File: deployment/proxy/Dockerfile (hunk 1 above)
Comment (skovati) [reply deployment/proxy/Dockerfile]: Good call, I'll make that change

File: deployment/proxy/docker-compose-proxy.yml (hunk 2 above)
Comment (skovati) [reply deployment/proxy/docker-compose-proxy.yml:21]: This is a newer feature of docker compose, that let's you merge together multiple `docker-compose.yml` files. The `!reset[]` directive tell the merge algorithm to completely override the ports section of our regular `deployment/docker-compose.yml`, i.e. it closes the previously exposed ports, since `nginx` handles them now.

https://docs.docker.com/compose/multiple-compose-files/merge/
